torch/csrc/jit/import.cpp

```diff
@@ -553,20 +222,225 @@ void reconstructOutputTypes(Block *b) {
+
+  virtual void buildIntermediateValue(Value* value, const std::string& name) override;
+
+  at::Tensor buildParameter(const onnx_torch::TensorProto& tensor_proto);
+
+  at::Tensor buildTensorCommon(const onnx_torch::TensorProto& tensor_proto,
+                               const int64_t storage_offset,
+                               const std::vector<int64_t>& strides);
+
+  std::pair<std::shared_ptr<script::Module>, std::string> parseFullName(
+      std::shared_ptr<script::Module> root_module,
+      const std::string fullname);
+
+  const std::unordered_map<std::string, std::string> *storage_export_map_;
+  std::unordered_map<std::string, std::shared_ptr<at::Tensor>> storage_map_;
+  std::unordered_map<std::string, const onnx_torch::TypeProto*> value_type_map_;
+};
+
+std::shared_ptr<Graph> ModuleDecoder::buildGraph(const onnx_torch::GraphProto& graph_proto) {
+  for (auto &subtype : graph_proto.value_info()) {
+    value_type_map_[subtype.name()] = &subtype.type();
+  }
+  return DecoderBase::buildGraph(graph_proto);
+}
+
+TypePtr ModuleDecoder::buildType(const onnx_torch::TypeProto& type_proto) {
+  auto tensortype_proto = type_proto.tensor_type();
+  auto shape_proto = tensortype_proto.shape();
+  auto kind = type_proto.denotation();
+  if (kind == "DynamicType") {
+    return DynamicType::get();
+  } else if (kind == "TensorType") {
+    // TODO: Don't use DynamicType here
+    return DynamicType::get();
+  } else if (kind == "TupleType") {
+    std::vector<TypePtr> elems;
+    for (auto &subkind : shape_proto.dim()) {
+      auto it = value_type_map_.find(subkind.dim_param());
+      JIT_ASSERT(it != value_type_map_.end());
+      elems.push_back(buildType(*it->second));
+    }
+    return TupleType::create(elems);
+  } else if (kind == "ListType") {
+    auto subkind = shape_proto.dim(0);
+    auto it = value_type_map_.find(subkind.dim_param());
+    JIT_ASSERT(it != value_type_map_.end());
+    return ListType::create(buildType(*it->second));
+  } else if (kind == "NumberType") {
+    return NumberType::get();
+  } else if (kind == "FloatType") {
+    return FloatType::get();
+  } else if (kind == "IntType") {
+    return IntType::get();
+  } else if (kind == "NoneType") {
+    return NoneType::get();
+  } else {
+    throw std::runtime_error("unexpected string for type kind");
+  }
+}
+
+void ModuleDecoder::buildValue(Value* value, const onnx_torch::ValueInfoProto& valueinfo_proto) {
+  value->setType(buildType(valueinfo_proto.type()));
+}
+
+void ModuleDecoder::buildIntermediateValue(Value* value, const std::string& name) {
+  auto it = value_type_map_.find(name);
+  JIT_ASSERT(it != value_type_map_.end());
+  value->setType(buildType(*it->second));
+}
+
+at::Tensor ModuleDecoder::buildParameter(const onnx_torch::TensorProto& tensor_proto) {
+  std::vector<int64_t> strides;
+  std::move(tensor_proto.int64_data().begin() + 3, tensor_proto.int64_data().end(), std::back_inserter(strides));
```

Comment:
Why are parameter tensors different than other tensors? We should at least have a comment. The +1 and the +3 are completely magical here.